File: relay/clock.py
from __future__ import annotations
from dataclasses import dataclass


DEFAULT_SCAN_PERIOD_MS = 10.0


@dataclass(frozen=True)
class SimClock:
    tick: int
    elapsed_ms: float

    def advance(self, scan_period_ms: float) -> SimClock:
        return SimClock(tick=self.tick + 1, elapsed_ms=self.elapsed_ms + scan_period_ms)

    @staticmethod
    def zero() -> SimClock:
        return SimClock(tick=0, elapsed_ms=0.0)

File: host/src/clock.cpp
#include "relay_host/clock.hpp"

namespace relay_host {

SimClock SimClock::advance(double scan_period_ms) const noexcept {
    return SimClock{tick + 1, elapsed_ms + scan_period_ms};
}

SimClock SimClock::zero() noexcept {
    return SimClock{0, 0.0};
}

}  // namespace relay_host
